cmd/freeze-checkpoint/app/root.go

```diff
@@ -0,0 +1,230 @@
+	case viper.GetString("signer-kmskey") != "":
+		kmshash := viper.GetString("signer-kmshash")
+		hashAlg, ok := hashAlgMap[kmshash]
+		if !ok {
+			return nil, fmt.Errorf("invalid hash algorithm for --signer-kmshash: %s", kmshash)
+		}
+		opts = []signerverifier.Option{signerverifier.WithKMS(viper.GetString("signer-kmskey"), hashAlg)}
+	case viper.GetString("signer-tink-kek-uri") != "":
+		opts = []signerverifier.Option{signerverifier.WithTink(viper.GetString("signer-tink-kek-uri"), viper.GetString("signer-tink-keyset-path"))}
+	default:
+		return nil, fmt.Errorf("must provide a signer using a file, KMS, or Tink")
+	}
+	signerVerifier, err := signerverifier.New(ctx, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("initializing key signer: %w", err)
+	}
+	return signerVerifier, nil
+}
+
+func getNoteSigner(ctx context.Context, signer signature.Signer) (note.Signer, error) {
+	origin := viper.GetString("hostname")
+	noteSigner, err := rekornote.NewNoteSigner(ctx, origin, signer)
+	if err != nil {
+		return nil, fmt.Errorf("initializing note signer: %w", err)
+	}
+	return noteSigner, nil
+}
+
+func getNoteVerifier(verifier signature.Verifier) (note.Verifier, error) {
+	origin := viper.GetString("hostname")
+	noteVerifier, err := rekornote.NewNoteVerifier(origin, verifier)
+	if err != nil {
+		return nil, fmt.Errorf("initializing note verifier: %w", err)
+	}
+	return noteVerifier, nil
+}
+
+func objectAccess(ctx context.Context) (*gcs.Reader, *gcs.Writer, error) {
+	client, err := gcs.NewClient(ctx, gcs.WithJSONReads())
+	if err != nil {
+		return nil, nil, fmt.Errorf("getting GCS client: %w", err)
+	}
+	bucketName := viper.GetString("gcp-bucket")
+	object := client.Bucket(bucketName).Object(layout.CheckpointPath)
+	objReader, err := object.NewReader(ctx)
+	if err != nil {
+		return nil, nil, fmt.Errorf("getting object reader: %w", err)
+	}
+	contentType := "text/plain; charset=utf-8"
+	cacheControl := "no-cache"
+	objWriter := object.NewWriter(ctx)
+	objWriter.ContentType = contentType
+	objWriter.CacheControl = cacheControl
+	return objReader, objWriter, nil
+}
+
+func getCheckpoint(objReader *gcs.Reader, noteVerifier note.Verifier) (*logformat.Checkpoint, error) {
+	rawCheckpoint, err := io.ReadAll(objReader)
+	if err != nil {
+		return nil, fmt.Errorf("reading object: %w", err)
+	}
+	noteObj, err := note.Open(rawCheckpoint, note.VerifierList(noteVerifier))
+	if err != nil {
+		return nil, fmt.Errorf("opening checkpoint: %w", err)
+	}
+	checkpoint := logformat.Checkpoint{}
+	rest, err := checkpoint.Unmarshal([]byte(noteObj.Text))
+	if err != nil {
+		return nil, fmt.Errorf("parsing checkpoint: %w", err)
+	}
+	if strings.Contains(string(rest), frozenString) {
+		return nil, nil
+	}
+	return &checkpoint, nil
+}
+
+func updateCheckpoint(objWriter *gcs.Writer, noteSigner note.Signer, checkpoint *logformat.Checkpoint) error {
+	buf := bytes.NewBuffer(checkpoint.Marshal())
```

Comment:
tiny nit, can you add a comment that this marshalled checkpoint is just the note and not the signatures? I double checked, wanted to make sure we aren't appending after the sigs